File: docs/source/_static/examples/cpp/ansatz.cpp
// Ansatz usage examples.

// --------------------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.
// --------------------------------------------------------------------------------------------

// --------------------------------------------------------------------------------------------
// start-cell-create
#include <iostream>
#include <qdk/chemistry.hpp>
#include <string>
using namespace qdk::chemistry::data;
using namespace qdk::chemistry::algorithms;

// create H2 structure
std::vector<Eigen::Vector3d> coords = {{0.0, 0.0, 0.0},
                                       {0.0, 0.0, 1.4}};  // Bohr
std::vector<std::string> symbols = {"H", "H"};

Structure structure(coords, symbols);

// SCF
auto scf_solver = ScfSolverFactory::create();
auto [E_scf, wfn_scf] = scf_solver->run(structure, 0, 1);

// Create hamiltonian from scf orbitals
auto ham_constructor = HamiltonianConstructorFactory::create();
auto hamiltonian = ham_constructor->run(wfn_scf->get_orbitals());

// Create ansatz
Ansatz ansatz(hamiltonian, wfn_scf);
// end-cell-create
// --------------------------------------------------------------------------------------------

// --------------------------------------------------------------------------------------------
// start-cell-access
// Access ansatz components
auto hamiltonian_ref = ansatz.get_hamiltonian();
auto wavefunction_ref = ansatz.get_wavefunction();
auto orbitals_ref = ansatz.get_orbitals();

// Check component availability
bool has_hamiltonian = ansatz.has_hamiltonian();
bool has_wavefunction = ansatz.has_wavefunction();
bool has_orbitals = ansatz.has_orbitals();

// Calculate energy expectation value
double energy = ansatz.calculate_energy();
std::cout << "Energy expectation value " << energy << " Ha" << std::endl;

// Get summary
std::string summary = ansatz.get_summary();
std::cout << "Ansatz summary " << summary << std::endl;
// end-cell-access
// --------------------------------------------------------------------------------------------

File: docs/source/_static/examples/python/ansatz.py
"""Basic Ansatz creation and usage examples."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

################################################################################
# docs:xyz ../data/h2.structure.xyz
# start-cell-create
from qdk_chemistry.data import Ansatz, Structure
from qdk_chemistry.algorithms import create

# Load H2 molecule structure from XYZ file
structure = Structure.from_xyz("""\
2
H2 molecule
H    0.000000    0.000000    0.000000
H    0.000000    0.000000    0.740848
""")

# SCF
scf_solver = create("scf_solver")
E_scf, wfn_scf = scf_solver.run(
    structure, charge=0, spin_multiplicity=1, basis_or_guess="sto-3g"
)

# Create Hamiltonian from SCF orbitals
hamiltonian_constructor = create("hamiltonian_constructor")
hamiltonian = hamiltonian_constructor.run(wfn_scf.get_orbitals())

# Create ansatz
ansatz_scf = Ansatz(hamiltonian, wfn_scf)
# end-cell-create
################################################################################

################################################################################
# start-cell-access
# Access ansatz components
hamiltonian_ref = ansatz_scf.get_hamiltonian()
wavefunction_ref = ansatz_scf.get_wavefunction()
orbitals_ref = ansatz_scf.get_orbitals()

# Check component availability
has_hamiltonian = ansatz_scf.has_hamiltonian()
has_wavefunction = ansatz_scf.has_wavefunction()
has_orbitals = ansatz_scf.has_orbitals()

# Calculate energy expectation value
energy = ansatz_scf.calculate_energy()
print(f"Energy expectation value: {energy:.8f} Hartree")

# Get summary information
summary = ansatz_scf.get_summary()
print(f"Ansatz summary: {summary}")
# end-cell-access
################################################################################
